### PR TITLE
texlive: use version info from tlpdb instead of hardcoding

### PR DESCRIPTION
**Pull request by xworld21**
###### Description of changes
Use the version information (TeX Live year, final vs snapshot, package database revision) extracted from `texlive.tlpdb.xz` instead of hardcoding most of it in `default.nix` and `bin.nix`.

This is almost a non-visible change, except that I have added the TLPDB revision number to the `texdoc` package, because its fixed output depends on the content of TLPDB too. This is necessary for the new fixed hashes workflow (which is more or less ready, to be submitted after this PR) where we recompile only the hashes of the *new* packages.

###### Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/typesetting/tex/texlive/tlVersion.nix [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  year = "2021";
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/tlVersion.nix:2]: First mistake: texlive year and snapshot year are not the same thing. So a better approach is to have attributes mirrors, bin, tlpdb, where bin and tlpdb have their own year/month/day.

**Comment by xworld21** [comment]
@veprbl this is now ready for review

**Review by apfelkuchen6 — COMMENTED** [review]
I like the idea. Using `tlpdb.nix` as the input for its own generation feels a little clumsy to me though. I think explicitly setting `frozen` (and `year` in case of the eol version) in `default.nix` and using that to generate `tlpdb.nix` would be preferable.

We could then check in `combine` whether the `frozen` and `year` values specified in `default.nix` and locked in `tlpdb.nix` match.

**Comment by xworld21** [comment]
> I think explicitly setting `frozen` (and `year` in case of the eol version) in `default.nix` and using that to generate `tlpdb.nix` would be preferable.

I agree, especially after I realised one has to set `frozen` manually to switch between snapshot and final, and that's cumbersome. We could assert they must agree either in `combine` or in `tests.texlive` (I think I prefer the latter which already has a test to check that `tlpdb.nix` is valid).

File: pkgs/tools/typesetting/tex/texlive/default.nix [hunk 2]
```diff
@@ -28,9 +28,14 @@ let
...
 
+  tlpdb = import ./tlpdb.nix;
+
+  tlpdbVersion = tlpdb."00texlive.config";
+
   # the set of TeX Live packages, collections, and schemes; using upstream naming
-  tl = let
-    orig = import ./tlpdb.nix;
+  tl = assert lib.assertMsg (tlpdbVersion.year == version.texliveYear) "TeX Live year in tlpdb.nix and texlive do not match, refusing to evaluate";
+    assert lib.assertMsg (tlpdbVersion.frozen == version.final) "TeX Live final status in tlpdb.nix and texlive do not match, refusing to evaluate"; let
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/default.nix:37]: Nix is lazy, but frustratingly not lazy enough: this assert makes it impossible to evaluate `texlive.tlpdb-nix`, even though it is independent of `tl`. Is there a better place for this assert?

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix]: Oof, I forgot that `texlive = tl / ...` so evaluating `texlive` itself calls the assert. Fixed.

**Comment by xworld21** [comment]
I acted on @apfelkuchen6's comment, now the expected version must be declared explicitly, and an `assert` prevents from evaluating `texlive.bin` and `texlive.combine(d?)` if there is a mismatch. I'd prefer that the packages themselves do not evaluate, but I am worried about adding performance issues, and the current approach is definitely good enough for OfBorg to stop a PR in its tracks right away.

File: pkgs/tools/typesetting/tex/texlive/UPGRADING.md [hunk 3]
```diff
@@ -1,62 +1,46 @@
...
-
-We mirror CTAN sources of this snapshot on community-operated servers and on IPFS.
-
-To upgrade the package snapshot, follow this process:
-
-
-### Snapshot sources and texlive package database
-
-Mirror the current CTAN archive to our mirror(s) and IPFS (URLs in `default.nix`).
-See https://tug.org/texlive/acquire-mirror.html for instructions.
```
Comment (apfelkuchen6) [inline pkgs/tools/typesetting/tex/texlive/UPGRADING.md:31]: I'm so glad this was removed. When I first read this document, this really confused me ("I'm supposed to do what?")

**Review by apfelkuchen6 — COMMENTED** [review]
This looks good to me now. We probably also need to target staging here (and for all other texlive changes) or wait until #222471 hits stable, right?

**Comment by xworld21** [comment]
> This looks good to me now. We probably also need to target staging here (and for all other texlive changes) or wait until #222471 hits stable, right?

If the po4a changes reduces the number of rebuilds substantially, I'd rather wait. Staging fails more often than master and makes it harder to trust we are not slipping in typos.

**Comment by apfelkuchen6** [comment]
> If the po4a changes reduces the number of rebuilds substantially, I'd rather wait. Staging fails more often than master and makes it harder to trust we are not slipping in typos.

After the po4a change, texlive changes should only cause [513 rebuilds](https://gist.github.com/GrahamcOfBorg/d0b1c8716ba2e1388734cecf332cf0c8), (~120 per platform), and at first glance most of them appear to be reasonable. Many of the remaining "unclear" ones appear to originate from PARI/GP, which generates its documentation with tex (which is unfortunate).

513 rebuilds in total should be good enough to target master directly, I think. #223303 has orange ofborg labels.

**Comment by xworld21** [comment]
Rebased against staging, and apologies to all the people I have just pinged by accident. Note that the tests used to pass on master.

This includes the fix for texlive's `extraVersion` which we forgot to change during the last update.

**Comment by xworld21** [comment]
Draft until #224806 is merged.

**Review by apfelkuchen6 — APPROVED** [review]

File: pkgs/tools/typesetting/tex/texlive/default.nix [hunk 4]
```diff
@@ -109,39 +115,46 @@ let
...
-    urls = map (up: "${up}/../tlpkg/texlive.tlpdb.xz") urlPrefixes;
+  tlpdbxz = fetchurl {
+    urls = map (up: "${up}/tlpkg/texlive.tlpdb.xz") urlPrefixes;
     hash = "sha256-vm7DmkH/h183pN+qt1p1wZ6peT2TcMk/ae0nCXsCoMw=";
   };
 
-  tlpdb-nix = runCommand "tlpdb.nix" {
-    inherit tlpdb;
+  tlpdbNix = runCommand "tlpdb.nix" {
+    inherit tlpdbxz;
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/default.nix:153]: Before calling this fully done: how do I hide this from [search.nixos.org](https://search.nixos.org/packages?channel=unstable&query=tlpdb)? It's ugly!

Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/default.nix:153]: There doesn't appear to be a pretty way to hide derivations.

the best solutions I can think of are
- nesting it inside another attr-set without `recurseForDerivations` attribute (like `pkgs.texlive.updater.tlpdbNix`). The derivations for the fixed hashes from the other PR could also be inserted there.
- moving it to the tests. After all, this is the only place where it is used aside from the manual invocation when updating. But reading something like „run `cp $(nix-build ../../../.. -A tests.texlive.tlpdbNix) tlpdb.nix`“ in the update instructions might also  cause confusion.

Also possible, but imho even worse:
- hiding it as the return value of a constant function.
- hiding it in a list
- hiding it as a passthru-attribute of one of the mostly unrelated derivations
- making it completely inaccessible via the pkgs datastructure (as before).

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:153]: Ok, I have opted for nesting under `texlive.tlpdb`, and while I was there, I exposed `tlpdbxz` and `version`.

File: pkgs/tools/typesetting/tex/texlive/default.nix [hunk 5]
```diff
@@ -198,9 +211,22 @@ let
...
+
 in
   tl // {
-    inherit bin combine tlpdb-nix;
+
+    tlpdb = {
+      # nested in an attribute set to prevent them from appearing in search
+      nix = tlpdbNix;
+      xz = tlpdbxz;
+      inherit version;
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/default.nix:225]: I am undecided about this line – should we or should we not expose the `version` attribute set? The details are somewhat available already from `extraVersion`.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:225]: Just as I ask the question, I realise it's better not to expose more than strictly necessary.

**Comment by xworld21** [comment]
@veprbl this is now ready for review and possibly merge.

**Review by apfelkuchen6 — APPROVED** [review]
LGTM. We should definitely get this in before 23.05 branchoff to sneakily fix the incorrect version number :D

**Comment by xworld21** [comment]
pinging @veprbl – it's probably fallen off his github screen (the first version was from December!)

**Review by veprbl — APPROVED** [review]

**Comment by veprbl** [comment]
@GrahamcOfBorg build tests.texlive

**Pull request merged**